### DISCUSSION TITLE
globalShortcut API: documentation for valid keys and modifiers?

### DISCUSSION THREAD

**Discussion by Sejmou** [tauri-apps/tauri#7121] [Q&A]
Tauri features an [API](https://tauri.app/v1/api/js/globalshortcut/) for defining global shortcuts via JavaScript. I managed to add it to the allowlist and register basic shortcuts for my application. However, I want to define more sophisticated shortcuts and would need some documentation for how to define the "shortcut command strings" to pass to the `register()` or `registerAll()` functions.

For reference, [Electron has documentation for the accepted key combination of its global shortcut feature](https://www.electronjs.org/docs/latest/api/accelerator), listing accepted key codes and modifiers.

Where can I find a list of accepted key codes/modifiers for Tauri? If no documentation exists yet, where should I look for the accepted values in the codebase?

Btw, I am aware that this question was already asked before (#2378 ), but it seems to have been ignored.

Thanks in advance for any help!

**Comment by Sejmou** [comment]
I am specifically looking for a way to bind the `MediaNextTrack`, `MediaPreviousTrack`, `MediaStop` and `MediaPlayPause` keys. So, if someone happens to know how to do that please let me know!

UPDATE: I found two files that are probably relevant for the shortcut logic:
https://github.com/tauri-apps/tao/blob/dev/src/keyboard.rs
https://github.com/tauri-apps/tao/blob/dev/src/accelerator.rs

Based on the code here, I would guess that
```typescript
await register('MEDIAPLAYPAUSE', () => {
  console.log('Shortcut triggered');
});
```

should work. But if I do that, the application fails to start (WebView window remains white), but for some reason no errors are logged. I am running the application via `tauri dev` on a M1 MacBook Pro with macOS 13.

If I change 'MEDIAPLAYPAUSE' to something invalid like 'MEDIAPLAYPAUS' I do get an error in the console though:

```rust
thread 'tokio-runtime-worker' panicked at 'invalid accelerator: AcceleratorParseError("Couldn't identify \"MEDIAPLAYPAUS\" as a valid `KeyCode`")', /Users/sejmou/.cargo/registry/src/github.com-1ecc6299db9ec823/tauri-runtime-wry-0.13.0/src/global_shortcut.rs:77:60
```

**Comment by ryanrixxh** [comment]
Having a similar issue with Numpad related Keycodes. When passing in any Numpad number like so: 

```
register('Numpad6', () => {
   console.log('Shortcut triggered')
})
```

 I receive this console error:
 `Uncaught (in promise) Unable to register accelerator (unknown VKCode for this char).`

Attempting to use a known VKCode like 'VK_NUMPAD6' removes the console error, but the console log is never triggered when pressing the hotkey.
 
 I'm unsure if this is in any way related to your problem, but the issue is that it is very hard to find that out, because this feature is mostly undocumented.

**Comment by Ar-mane** [comment]
i was invetigating this, found that it uses [this set ](https://github.com/tauri-apps/global-hotkey/blob/c9913a97667b3e44cb000de384cd8937d5a0050a/src/hotkey.rs#L212)

parsing your shortcut 

`fn parse_hotkey(hotkey: &str) -> Result<HotKey, HotKeyParseError> ` 

so you can easily use this as reference

**Reply by Ar-mane** [reply]
```ts
export const keyToCodeMap = {
  Shift: 'SHIFT',
  Control: 'CONTROL',
  Alt: 'ALT',
  Super: 'SUPER',
  '`': 'BACKQUOTE',
  '\\': 'BACKSLASH',
  '[': 'BRACKETLEFT',
  ']': 'BRACKETRIGHT',
  ',': 'COMMA',
  '0': 'DIGIT0',
  '1': 'DIGIT1',
  '2': 'DIGIT2',
  '3': 'DIGIT3',
  '4': 'DIGIT4',
  '5': 'DIGIT5',
  '6': 'DIGIT6',
  '7': 'DIGIT7',
  '8': 'DIGIT8',
  '9': 'DIGIT9',
  '=': 'EQUAL',
  A: 'KEYA',
  B: 'KEYB',
  C: 'KEYC',
  D: 'KEYD',
  E: 'KEYE',
  F: 'KEYF',
  G: 'KEYG',
  H: 'KEYH',
  I: 'KEYI',
  J: 'KEYJ',
  K: 'KEYK',
  L: 'KEYL',
  M: 'KEYM',
  N: 'KEYN',
  O: 'KEYO',
  P: 'KEYP',
  Q: 'KEYQ',
  R: 'KEYR',
  S: 'KEYS',
  T: 'KEYT',
  U: 'KEYU',
  V: 'KEYV',
  W: 'KEYW',
  X: 'KEYX',
  Y: 'KEYY',
  Z: 'KEYZ',
  '-': 'MINUS',
  '.': 'PERIOD',
  "'": 'QUOTE',
  ';': 'SEMICOLON',
  '/': 'SLASH',
  Backspace: 'BACKSPACE',
  CapsLock: 'CAPSLOCK',
  Enter: 'ENTER',
  ' ': 'SPACE',
  Tab: 'TAB',
  Delete: 'DELETE',
  End: 'END',
  Home: 'HOME',
  Insert: 'INSERT',
  PageDown: 'PAGEDOWN',
  PageUp: 'PAGEUP',
  PrintScreen: 'PRINTSCREEN',
  ScrollLock: 'SCROLLLOCK',
  ArrowDown: 'ARROWDOWN',
  ArrowLeft: 'ARROWLEFT',
  ArrowRight: 'ARROWRIGHT',
  ArrowUp: 'ARROWUP',
  NumLock: 'NUMLOCK',
  Numpad0: 'NUMPAD0',
  Numpad1: 'NUMPAD1',
  Numpad2: 'NUMPAD2',
  Numpad3: 'NUMPAD3',
  Numpad4: 'NUMPAD4',
  Numpad5: 'NUMPAD5',
  Numpad6: 'NUMPAD6',
  Numpad7: 'NUMPAD7',
  Numpad8: 'NUMPAD8',
  Numpad9: 'NUMPAD9',
  NumpadAdd: 'NUMPADADD',
  NumpadDecimal: 'NUMPADDECIMAL',
  NumpadDivide: 'NUMPADDIVIDE',
  NumpadEnter: 'NUMPADENTER',
  NumpadEqual: 'NUMPADEQUAL',
  NumpadMultiply: 'NUMPADMULTIPLY',
  NumpadSubtract: 'NUMPADSUBTRACT',
  Escape: 'ESCAPE',
  F1: 'F1',
  F2: 'F2',
  F3: 'F3',
  F4: 'F4',
  F5: 'F5',
  F6: 'F6',
  F7: 'F7',
  F8: 'F8',
  F9: 'F9',
  F10: 'F10',
  F11: 'F11',
  F12: 'F12',
  AudioVolumeDown: 'AUDIOVOLUMEDOWN',
  AudioVolumeUp: 'AUDIOVOLUMEUP',
  AudioVolumeMute: 'AUDIOVOLUMEMUTE',
  MediaPlay: 'MEDIAPLAY',
  MediaPause: 'MEDIAPAUSE',
  MediaPlayPause: 'MEDIAPLAYPAUSE',
  MediaStop: 'MEDIASTOP',
  MediaTrackNext: 'MEDIATRACKNEXT',
  MediaTrackPrevious: 'MEDIATRACKPREV',
  F13: 'F13',
  F14: 'F14',
  F15: 'F15',
  F16: 'F16',
  F17: 'F17',
  F18: 'F18',
  F19: 'F19',
  F20: 'F20',
  F21: 'F21',
  F22: 'F22',
  F23: 'F23',
  F24: 'F24'
};
```
